File: src/amodhakal/lox/Scanner.java
package amodhakal.lox;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Scanner {
   private final String source;
   private Integer start = 0;
   private Integer current = 0;
   private Integer line = 1;

   private final List<Token> tokens;
   private static final Map<String, TokenType> keywords;

   static {
      keywords = new HashMap<>();
      keywords.put("and", TokenType.AND);
      keywords.put("class", TokenType.CLASS);
      keywords.put("else", TokenType.ELSE);
      keywords.put("false", TokenType.FALSE);
      keywords.put("for", TokenType.FOR);
      keywords.put("fun", TokenType.FUN);
      keywords.put("if", TokenType.IF);
      keywords.put("nil", TokenType.NIL);
      keywords.put("or", TokenType.OR);
      keywords.put("print", TokenType.PRINT);
      keywords.put("return", TokenType.RETURN);
      keywords.put("super", TokenType.SUPER);
      keywords.put("this", TokenType.THIS);
      keywords.put("true", TokenType.TRUE);
      keywords.put("var", TokenType.VAR);
      keywords.put("while", TokenType.WHILE);
   }

   public Scanner(String source) {
      this.source = source;
      tokens = new ArrayList<>();
   }

   public List<Token> scanTokens() {
      while (!isAtEnd()) {
         start = current;
         scanToken();
      }

      tokens.add(new Token(TokenType.EOF, "", null, line));
      return tokens;
   }

   private Boolean isAtEnd() {
      return current >= source.length();
   }

   private void scanToken() {
      Character ch = advance();
      switch (ch) {
         case '(':
            addToken(TokenType.LEFT_PAREN);
            break;
         case ')':
            addToken(TokenType.RIGHT_PAREN);
            break;
         case '{':
            addToken(TokenType.LEFT_BRACE);
            break;
         case '}':
            addToken(TokenType.RIGHT_BRACE);
            break;
         case ',':
            addToken(TokenType.COMMA);
            break;
         case '.':
            addToken(TokenType.DOT);
            break;
         case '-':
            addToken(TokenType.MINUS);
            break;
         case '+':
            addToken(TokenType.PLUS);
            break;
         case ';':
            addToken(TokenType.SEMICOLON);
            break;
         case '*':
            addToken(TokenType.STAR);
            break;
         case '!':
            addToken(match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
            break;
         case '=':
            addToken(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
            break;
         case '<':
            addToken(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
            break;
         case '>':
            addToken(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
            break;
         case '/':
            if (match('/')) {
               while (peek() != '\n' && !isAtEnd()) advance();
            } else {
               addToken(TokenType.SLASH);
            }
            break;
         case ' ':
         case '\r':
         case '\t':
            break;
         case '\n':
            line++;
            break;
         case '"':
            string();
            break;
         case 'o':
            if (match('r')) {
               addToken(TokenType.OR);
            }
            break;
         default:
            if (isDigit(ch)) {
               number();
            } else if (isAlpha(ch)) {
               identifier();
            } else {
               Lox.error(line, "Unexpected character.");
            }
            break;
      }
   }

   private void identifier() {
      while (isAlphaNumeric(peek())) advance();
      String text = source.substring(start, current);
      TokenType type = keywords.get(text);
      if (type == null) type = TokenType.IDENTIFIER;
      addToken(type);
   }

   private void number() {
      while (isDigit(peek())) advance();

      if (peek() == '.' && isDigit(peekNext())) {
         do advance();
         while (isDigit(peek()));
      }

      addToken(TokenType.NUMBER, Double.parseDouble(source.substring(start, current)));
   }

   private void string() {
      while (peek() != '"' && !isAtEnd()) {
         if (peek() == '\n') line++;
         advance();
      }

      if (isAtEnd()) {
         Lox.error(line, "Unterminated string");
         return;
      }

      advance();
      String value = source.substring(start + 1, current - 1);
      addToken(TokenType.STRING, value);
   }

   private Character advance() {
      return source.charAt(current++);
   }

   private void addToken(TokenType type) {
      addToken(type, null);
   }

   private void addToken(TokenType type, Object literal) {
      String text = source.substring(start, current);
      tokens.add(new Token(type, text, literal, line));
   }

   private Boolean match(Character expected) {
      if (isAtEnd()) return false;
      if (source.charAt(current) != expected) return false;

      current++;
      return true;
   }

   private Character peek() {
      if (isAtEnd()) return '\0';
      return source.charAt(current);
   }

   private Character peekNext() {
      if (current + 1 >= source.length()) return '\0';
      return source.charAt(current + 1);
   }

   private Boolean isAlphaNumeric(Character ch) {
      return isAlpha(ch) || isDigit(ch);
   }

   private Boolean isAlpha(Character ch) {
      return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch == '_';
   }

   private Boolean isDigit(Character ch) {
      return ch >= '0' && ch <= '9';
   }
}

File: src/amodhakal/lox/Scanner.kt
package amodhakal.lox

import com.sun.org.apache.bcel.internal.generic.RETURN
import com.sun.tools.example.debug.expr.ExpressionParserConstants.ELSE
import com.sun.tools.example.debug.expr.ExpressionParserConstants.FOR
import com.sun.tools.example.debug.expr.ExpressionParserConstants.IF
import com.sun.tools.example.debug.expr.ExpressionParserConstants.WHILE
import com.sun.tools.javac.code.Kinds.KindSelector.NIL
import java.util.HashMap


class Scanner internal constructor(private val source: String) {
    private val tokens: MutableList<Token> = ArrayList()
    private var start = 0
    private var current = 0
    private var line = 1

    private var keywords: MutableMap<String, TokenType> = initKeywords()

    private fun initKeywords(): MutableMap<String, TokenType> {
        val keywords = HashMap<String, TokenType>()
        keywords["and"] = TokenType.AND;
        keywords["class"] = TokenType.CLASS;
        keywords["else"] = TokenType.ELSE;
        keywords["false"] = TokenType.FALSE;
        keywords["for"] = TokenType.FOR;
        keywords["fun"] = TokenType.FUN;
        keywords["if"] = TokenType.IF;
        keywords["nil"] = TokenType.NIL;
        keywords["or"] = TokenType.OR;
        keywords["print"] = TokenType.PRINT;
        keywords["return"] = TokenType.RETURN;
        keywords["super"] = TokenType.SUPER;
        keywords["this"] = TokenType.THIS;
        keywords["true"] = TokenType.TRUE;
        keywords["var"] = TokenType.VAR;
        keywords["while"] = TokenType.WHILE;
        return keywords
    }


    fun scanTokens(): List<Token> {
        while (!isAtEnd) {
            start = current
            scanToken()
        }

        tokens.add(Token(TokenType.EOF, "", null, line))
        return tokens
    }

    private val isAtEnd: Boolean
        get() = current >= source.length

    private fun scanToken() {
        when (val ch = advance()) {
            '(' -> addToken(TokenType.LEFT_PAREN)
            ')' -> addToken(TokenType.RIGHT_PAREN)
            '{' -> addToken(TokenType.LEFT_BRACE)
            '}' -> addToken(TokenType.RIGHT_BRACE)
            ',' -> addToken(TokenType.COMMA)
            '.' -> addToken(TokenType.DOT)
            '-' -> addToken(TokenType.MINUS)
            '+' -> addToken(TokenType.PLUS)
            ';' -> addToken(TokenType.SEMICOLON)
            '*' -> addToken(TokenType.STAR)
            '!' -> addToken(if (match('=')) TokenType.BANG_EQUAL else TokenType.BANG)
            '=' -> addToken(if (match('=')) TokenType.EQUAL_EQUAL else TokenType.EQUAL)
            '>' -> addToken(if (match('=')) TokenType.GREATER_EQUAL else TokenType.GREATER)
            '<' -> addToken(if (match('=')) TokenType.LESS_EQUAL else TokenType.LESS)

            ' ', '\r', '\t' -> return
            '\n' -> line++

            '"' -> string()

            '/' -> if (match('/')) {
                while (peek() != '\n' && !isAtEnd) advance()
            } else {
                addToken(TokenType.SLASH)
            }

            else -> if (isDigit(ch)) {
                number()
            } else if (isAlpha(ch)) {
                identifier()
            } else {
                Lox.error(line, "Unexpected character.")
            }
        }
    }

    private fun identifier() {
        while (isAlphaNumeric(peek())) advance()
        val text = source.substring(start, current)
        var type = keywords[text]
        if (type == null) type = TokenType.IDENTIFIER
        addToken(type)
    }

    private fun number() {
        while (isDigit(peek())) advance()

        if (peek() == '.' && isDigit(peekNext())) {
            advance()
            while (isDigit(peek())) advance()
        }

        addToken(TokenType.NUMBER, source.substring(start, current).toDouble())
    }

    private fun string() {
        while (peek() != '"' && isAtEnd) {
            if (peek() == '\n') line++
            advance()
        }

        if (isAtEnd) {
            Lox.error(line, "Unterminated string")
            return
        }

        advance()
        val value = source.substring(start + 1, current - 1)
        addToken(TokenType.STRING, value)
    }

    private fun advance(): Char {
        return source[current++]
    }

    private fun addToken(type: TokenType) {
        addToken(type, null)
    }

    private fun addToken(type: TokenType, literal: Any?) {
        val text = source.substring(start, current)
        tokens.add(Token(type, text, literal, line))
    }

    private fun match(expected: Char): Boolean {
        if (isAtEnd) return false
        if (source[current] != expected) return false

        current++
        return true
    }

    private fun peek(): Char {
        if (isAtEnd) return '\u0000'
        return source[current]
    }

    private fun peekNext(): Char {
        if (current + 1 >= source.length) return '\u0000'
        return source[current + 1]
    }

    private fun isAlpha(ch: Char): Boolean {
        return (ch in 'a'..'z') || (ch in 'A'..'Z') || ch == '_'
    }

    private fun isAlphaNumeric(ch: Char): Boolean {
        return isAlpha(ch) || isDigit(ch)
    }

    private fun isDigit(ch: Char): Boolean {
        return ch in '0'..'9'
    }
}
